### ISSUE TITLE
gradio 4 critical bug -- all return messages truncated to 65k

### ISSUE THREAD

**Issue by pseudotensor** [gradio-app/gradio#6601]
### Describe the bug

In gradio 3 there were never any issues with returning large amounts of text or data.  However, in gradio 4 this is totally broken.

This is super critical bug given how broadly it applies to all API calls of any return type (text, audio, etc.).

Related: https://github.com/gradio-app/gradio/issues/6319

that is, even once heartbeat bug is fixed, long output still hits json error mentioned there.

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

server:
```
import gradio as gr
import random
import time

with gr.Blocks() as demo:
    chatbot = gr.Chatbot()
    msg = gr.Textbox()
    clear = gr.Button("Clear")

    def user(user_message, history):
        return "", history + [[user_message, None]]

    def bot(history):
        bot_message = ' '.join(['a'] * 100000)
        history[-1][1] = bot_message
        yield history

    msg.submit(user, [msg, chatbot], [msg, chatbot], queue=False).then(
        fn=bot, inputs=chatbot, outputs=chatbot, api_name='bot',
    )
    clear.click(lambda: None, None, chatbot, api_name='clear')

demo.queue()
demo.launch()
```

client:
```
import time
from gradio_client import Client

client = Client('http://localhost:7860', serialize=False)

args = [[['Who are you?', None]]]
res = client.predict(*tuple(args), api_name='/bot')
print(res)
```

Error:

```
(/data/conda/h2ogpt) jon@pseudotensor:~/h2ogpt$ python testchat_nostream_client.py 
Loaded as API: http://localhost:7860/ ✔
Traceback (most recent call last):
  File "/home/jon/h2ogpt/testchat_nostream_client.py", line 7, in <module>
    res = client.predict(*tuple(args), api_name='/bot')
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 305, in predict
    return self.submit(*args, api_name=api_name, fn_index=fn_index).result()
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 1456, in result
    return super().result(timeout=timeout)
  File "/data/conda/h2ogpt/lib/python3.10/concurrent/futures/_base.py", line 445, in result
    return self.__get_result()
  File "/data/conda/h2ogpt/lib/python3.10/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/data/conda/h2ogpt/lib/python3.10/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 869, in _inner
    predictions = _predict(*data)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 894, in _predict
    result = utils.synchronize_async(self._sse_fn, data, hash_data, helper)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/utils.py", line 665, in synchronize_async
    return fsspec.asyn.sync(fsspec.asyn.get_loop(), func, *args, **kwargs)  # type: ignore
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fsspec/asyn.py", line 103, in sync
    raise return_result
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fsspec/asyn.py", line 56, in _runner
    result[0] = await coro
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 1075, in _sse_fn
    return await utils.get_pred_from_sse(
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/utils.py", line 342, in get_pred_from_sse
    return task.result()
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/utils.py", line 374, in stream_sse
    resp = json.loads(line[5:])
  File "/data/conda/h2ogpt/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/data/conda/h2ogpt/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/data/conda/h2ogpt/lib/python3.10/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Unterminated string starting at: line 1 column 70 (char 69)
```

When checking the details, the issue is that all messages of any kind are truncated to no more than 65k bytes.  This was never a problem with gradio 3 and I can see large messages work perfectly fine.

For audio and video this is a DOA for gradio 4 and its API.

### System Info

```shell
gradio==4.7.1
gradio_client==0.7.0
```

### Severity

Blocking usage of gradio

**Comment by pseudotensor** [comment]
If It try to debug the clent for the above running server, and step through stream_sse(), eventually for no obvious reason it hits on client:

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/routing.py", line 274, in app
    raw_response = await run_endpoint_function(
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio/routes.py", line 668, in queue_data
    blocks._queue.attach_data(body)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio/queueing.py", line 161, in attach_data
    raise ValueError("Event not found", event_id)
ValueError: ('Event not found', 'c4d06de1de954425bc4fe30e64eee7fe')
```

and on server:
```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/routing.py", line 274, in app
    raw_response = await run_endpoint_function(
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio/routes.py", line 668, in queue_data
    blocks._queue.attach_data(body)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio/queueing.py", line 161, in attach_data
    raise ValueError("Event not found", event_id)
ValueError: ('Event not found', 'c4d06de1de954425bc4fe30e64eee7fe')
```

It's like the client-server connection is unstable to timing choices that come from how the async stream is handled.  That's not good.

**Comment by pseudotensor** [comment]
My guess is this is the offending changes: https://github.com/gradio-app/gradio/pull/6069

FYI @aliabid94

**Comment by pseudotensor** [comment]
Actually I can't tell where things changed.  Maybe make_predict changes by @pngwn

**Comment by pseudotensor** [comment]
If I start messing with the stream_sse() code, e.g. just adding a print or something, I can see random changes in behavior.  Sometimes the print shows full correct output, sometimes not.  All a mess.

i.e. just this debug:
```
        async with client.stream(
            "GET", sse_url, params=hash_data, cookies=cookies
        ) as response:
            async for line in response.aiter_text():
                if len(line) > 65000:
                    print(line, flush=True)
                    continue
                if line.startswith("data:"):

```

Not intended to make work, just shows that the printed line is changing each time one runs the client for  fixed server.

I cannot tell if server code or client code is wrong.

**Comment by pseudotensor** [comment]
This seems to work:
```
            async for line in response.aiter_lines():
                print(len(line), flush=True)
                if len(line) == 0:
                    continue
                if line.startswith("data:"):
```

i.e. aiter_lines() instead of aiter_text() and ignoring 0 length lines.

Related: https://github.com/encode/httpx/discussions/2310

**Comment by pseudotensor** [comment]
But it's not a perfect fix. I sometimes see:

```

Traceback (most recent call last):
  File "/home/jon/h2ogpt/gradio_utils/grclient.py", line 264, in submit
    self.refresh_client_if_should()
  File "/home/jon/h2ogpt/gradio_utils/grclient.py", line 210, in refresh_client_if_should
    server_hash = self.get_server_hash()
  File "/home/jon/h2ogpt/gradio_utils/grclient.py", line 203, in get_server_hash
    return super().submit(api_name="/system_hash").result()
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 1456, in result
    return super().result(timeout=timeout)
  File "/data/conda/h2ogpt/lib/python3.10/concurrent/futures/_base.py", line 445, in result
    return self.__get_result()
  File "/data/conda/h2ogpt/lib/python3.10/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/data/conda/h2ogpt/lib/python3.10/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 869, in _inner
    predictions = _predict(*data)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 894, in _predict
    result = utils.synchronize_async(self._sse_fn, data, hash_data, helper)
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/utils.py", line 667, in synchronize_async
    return fsspec.asyn.sync(fsspec.asyn.get_loop(), func, *args, **kwargs)  # type: ignore
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fsspec/asyn.py", line 103, in sync
    raise return_result
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/fsspec/asyn.py", line 56, in _runner
    result[0] = await coro
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/client.py", line 1075, in _sse_fn
    return await utils.get_pred_from_sse(
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/utils.py", line 342, in get_pred_from_sse
    return task.result()
  File "/data/conda/h2ogpt/lib/python3.10/site-packages/gradio_client/utils.py", line 413, in stream_sse
    raise ValueError(f"Unexpected message: {line}")
ValueError: Unexpected message: {"detail":"Not Found"}
```

**Comment by pseudotensor** [comment]
Also the "work-around" leads to stuff like this in response:

```
['[', "'", '', '', 'T', 'h', 'e', '', 'p', 'a', 'p', 'e', 'r', '', 'd', 'i', 's', 'c', 'u', 's', 's', 'e', 's', '', 't', 'h', 'e', '', 'd', 'e', 'v', 'e', 'l', 'o', 'p', 'm', 'e', 'n', 't', '', 'o', 'f', '', 'a', '', 's', 'p', 'e', 'e', 'c', 'h', '', 'r', 'e', 'c', 'o', 'g', 'n', 'i', 't', 'i', 'o', 'n', '', 's', 'y', 's', 't', 'e', 'm', '', 'c', 'a', 'l', 'l', 'e', 'd', '', 'W', 'h', 'i', 's', 'p', 'e', 'r', ',', '', 'w', 'h', 'i', 'c', 'h', '', 'i', 's', '', 't', 'r', 'a', 'i', 'n', 'e', 'd', '', 'o', 'n', '', 'a', '', 'l', 'a', 'r', 'g', 'e', '', 'd', 'a', 't', 'a', 's', 'e', 't', '', 'o', 'f', '', '6', '8', '0', ',', '0', '0', '0', '', 'h', 'o', 'u', 'r', 's', '', 'o', 'f', '', 'a', 'u', 'd', 'i', 'o', '', 'a', 'n', 'd', '', 'a', 'c', 'h', 'i', 'e', 'v', 'e', 's', '', 'h', 'i', 'g', 'h', '', 'a', 'c', 'c', 'u', 'r', 'a', 'c', 'y', '', 'w', 'i', 't', 'h', 'o', 'u', 't', '', 't', 'h', 'e', '', 'n', 'e', 'e', 'd', '', 'f', 'o', 'r', '', 'f', 'i', 'n', 'e', '-', 't', 'u', 'n', 'i', 'n', 'g', '.', '', 'T', 'h', 'e', '', 'a', 'u', 't', 'h', 'o', 'r', 's', '', 'a', 'r', 'g', 'u', 'e', '', 't', 'h', 'a', 't', '', 'e', 'x', 'i', 's', 't', 'i', 'n', 'g', '', 's', 'p', 'e', 'e', 'c', 'h', '', 'r', 'e', 'c', 'o', 'g', 'n', 'i', 't', 'i', 'o', 'n', '', 's', 'y', 's', 't', 'e', 'm', 's', '', 'a', 'r', 'e', '', 'l', 'i', 'm', 'i', 't', 'e', 'd', '', 'b', 'y', '', 't', 'h', 'e', 'i', 'r', '', 'r', 'e', 'l', 'i', 'a', 'n', 'c', 'e', '', 'o', 'n', '', 'h', 'u', 'm', 'a', 'n', '-', 'l', 'a', 'b', 'e', 'l', 'e', 'd', '', 'd', 'a', 't', 'a', ',', '', 'a', 'n', 'd', '', 't', 'h', 'a', 't', '', 't', 'h', 'e', 'i', 'r', '', 'a', 'p', 'p', 'r', 'o', 'a', 'c', 'h', ',', '', 'w', 'h', 'i', 'c', 'h', '', 'u', 's', 'e', 's', '', 'w', 'e', 'a', 'k', '', 's', 'u', 'p', 'e', 'r', 'v', 'i', 's', 'i', 'o', 'n', '', 'a', 'n', 'd', '', 'a', '', 'l', 'a', 'r', 'g', 'e', '', 'a', 'm', 'o', 'u', 'n', 't', '', 'o', 'f', '', 'u', 'n', 'l', 'a', 'b', 'e', 'l', 'e', 'd', '', 'd', 'a', 't', 'a', ',', '', 'i', 's', '', 'm', 'o', 'r', 'e', '', 'e', 'f', 'f', 'e', 'c', 't', 'i', 'v', 'e', '', 'a', 'n', 'd', '', 'r', 'o', 'b', 'u', 's', 't', '.', '', 'T', 'h', 'e', 'y', '', 'a', 'l', 's', 'o', '', 'e', 'm', 'p', 'h', 'a', 's', 'i', 'z', 'e', '', 't', 'h', 'e', '', 'i', 'm', 'p', 'o', 'r', 't', 'a', 'n', 'c', 'e', '', 'o', 'f', '', 'm', 'u', 'l', 't', 'i', 'l', 'i', 'n', 'g', 'u', 'a', 'l', '', 'a', 'n', 'd', '', 'm', 'u', 'l', 't', 'i', 't', 'a', 's', 'k', '', 't', 'r', 'a', 'i', 'n', 'i', 'n', 'g', ',', '', 'a', 'n', 'd', '', 'r', 'e', 'l', 'e', 'a', 's', 'e', '', 'i', 'n', 'f', 'e', 'r', 'e', 'n', 'c', 'e', '', 'c', 'o', 'd', 'e', '', 'a', 'n', 'd', '', 'm', 'o', 'd', 'e', 'l', 's', '', 't', 'o', '', 's', 'e', 'r', 'v', 'e', '', 'a', 's', '', 'a', '', 'f', 'o', 'u', 'n', 'd', 'a', 't', 'i', 'o', 'n', '', 'f', 'o', 'r', '', 'f', 'u', 'r', 't', 'h', 'e', 'r', '', 'r', 'e', 's', 'e', 'a', 'r', 'c', 'h', '.', '\\', 'n', '\\', 'n', 'T', 'h', 'e', '', 'a', 'u', 't', 'h', 'o', 'r', 's', '', 'm', 'e', 'n', 't', 'i', 'o', 'n', '', 'r', 'e', 'l', 'a', 't', 'e', 'd', '', 'w', 'o', 'r', 'k', '', 'i', 'n', '', 'c', 'o', 'm', 'p', 'u', 't', 'e', 'r', '', 'v', 'i', 's', 'i', 'o', 'n', ',', '', 'w', 'h', 'e', 'r', 'e', '', 'l', 'a', 'r', 'g', 'e', '', 'd', 'a', 't', 'a', 's', 'e', 't', 's', '', 'w', 'i', 't', 'h', '', 'w', 'e', 'a', 'k', '', 's', 'u', 'p', 'e', 'r', 'v', 'i', 's', 'i', 'o', 'n', '', 'h', 'a', 'v', 'e', '', 'b', 'e', 'e', 'n', '', 's', 'h', 'o', 'w', 'n', '', 't', 'o', '', 'i', 'm', 'p', 'r', 'o', 'v', 'e', '', 't', 'h', 'e', '', 'r', 'o', 'b', 'u', 's', 't', 'n', 'e', 's', 's', '', 'a', 'n', 'd', '', 'g', 'e', 'n', 'e', 'r', 'a', 'l', 'i', 'z', 'a', 't', 'i', 'o', 'n', '', 'o', 'f', '', 'm', 'o', 'd', 'e', 'l', 's', '.', '', 'T', 'h', 'e', 'y', '', 'a', 'l', 's', 'o', '', 'r', 'e', 'f', 
```

**Comment by abidlabs** [comment]
Thanks for the detailed report @pseudotensor. Is this only an issue when you are using the Client to make a prediction, or even if you use the Gradio app via UI?

**Comment by pseudotensor** [comment]
So far only seen in API use, and the only changes to work-around so far are in client side.

**Comment by pseudotensor** [comment]
Will probably have to revert back to gradio 3.  Just too unstable.

**Comment by abidlabs** [comment]
We're working through a few issues related to the Client, e.g. https://github.com/gradio-app/gradio/pull/6602

Btw was just looking at your Client code, and noticed that you have:

```py
import time
from gradio_client import Client

client = Client('http://localhost:7860', serialize=False)
```

In Gradio 4.x, you don't need to set `serialize=False`, since the Chatbot component now returns a list of tuples by default. You can just do:

 ```py
import time
from gradio_client import Client

client = Client('http://localhost:7860')
```

Will work through the other issues you mentioned here soon.

**Comment by pseudotensor** [comment]
I need sanitize=False in general for many other reasons.  E.g. if through API I push http link into text box, it gets converted into temp file on server.  I don't want any of those conversions done.

**Comment by abidlabs** [comment]
Got it okay, we can explore those issues later, let's get this unblocked first

**Comment by freddyaboulton** [comment]
Hi @pseudotensor ! Thanks for all of the helpful comments you left on this thread as you investigated the issue. I think I have a fix in #6693. I would appreciate if you could test it out. You can install the client from that PR with 

```
pip install "gradio-client @ git+https://github.com/gradio-
app/gradio@6887fe4e080647f9892b478c35a625b056eddb31#subdirectory=client/python"
```

That PR only targets the 65k response length issue and the hearbeat issue (#6319). There are still some other issues with the client we are fixing. And #6556 should hopefully fix a lot of them.
